bazel/deps-cache.sh

```diff
@@ -0,0 +1,203 @@
+#!/bin/bash
+#
+# Copyright 2023 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+# Usage:
+#   $ deps-cache.sh [-c] [<google_cloud_cpp_deps.bzl>]
+#   $ deps-cache.sh -p [<google_cloud_cpp_deps.bzl>]
+#
+#   Args:
+#     google_cloud_cpp_deps.bzl    Path to the .bzl file that specifies the
+#                                  HTTP archives that we cache in GCS. By
+#                                  default uses "google_cloud_cpp_deps.bzl"
+#                                  in the same directory as this script.
+#
+#   Options:
+#     -c     Check that the HTTP archives are correctly cached (the default)
+#     -p     Populate the cache with current copies of the HTTP archives
+#     -h     Print help message
+#
+# Note:
+#   This does not edit the .bzl file to add cache files as URLs. It only
+#   checks/populates the cached files themselves, so they're ready for use.
+
+set -euo pipefail
+
+# Extracts all the documentation at the top of this file as the usage text.
+USAGE="$(sed -n '/^# Usage:/,/^$/s/^# \?//p' "$0")"
+readonly USAGE
+
+declare -i POPULATE_FLAG=0
+while getopts "chp" opt "$@"; do
+  case "${opt}" in
+    c)
+      POPULATE_FLAG=
+      ;;
+    p)
+      POPULATE_FLAG=1
+      ;;
+    h)
+      echo "${USAGE}"
+      exit 0
+      ;;
+    *)
+      echo "${USAGE}"
+      exit 1
+      ;;
+  esac
+done
+shift $((OPTIND - 1))
+readonly POPULATE_FLAG
+
+BZL_FILE=
+case $# in
+  0)
+    BZL_FILE="$(dirname "$0")/google_cloud_cpp_deps.bzl"
+    ;;
+  1)
+    BZL_FILE="$1"
+    shift
+    ;;
+  *)
+    echo "${USAGE}"
+    exit 1
+    ;;
+esac
+readonly BZL_FILE
+
+PROGRAM="$(basename "$0" .sh)"
+TMPDIR=$(mktemp -d --tmpdir "${PROGRAM}.XXXXXXXXXX")
+readonly PROGRAM TMPDIR
+trap 'rm -fr "${TMPDIR}"' EXIT
+
+# Extract the http_archive information from ${BZL_FILE}.
+cat >"${TMPDIR}/script" <<'EOT'
```

Comment:
This is pretty impressive, and a little over my head.  If it helps (and only if it does), you can use something like this:

https://github.com/googleapis/google-cloud-cpp/compare/main...coryan:google-cloud-cpp:ci-automate-fallbacks-for-bazel-deps?expand=1

to extract the information.  Basically exploits the fact that a `.bzl` file has Python syntax.  If you stub out the functions you don't care about, and replace the functions you care about with things you want, you can get the output in an easier to parse.

Reply:
I had thought about something like that, but stuck with the devil I knew (and also given our penchant for `sed`ing).  But I'll re-examine given your example, and see what I can come up with.  Thanks.  Although, if we go that route, we might as well just do the whole thing in python.

Reply:
SGTM

Reply:
OK.  I'll merge as-is, and we can deal with any Python migration at a later date.